frontend/src/v5/ui/controls/formModal/formModal.component.tsx

```diff
@@ -0,0 +1,90 @@
+/**
+ *  Copyright (C) 2023 3D Repo Ltd
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { DetailedHTMLProps, FormHTMLAttributes } from 'react';
+import { Dialog } from '@mui/material';
+import { DialogProps } from '@mui/material/Dialog';
+import { Form, RemoveWhiteCorners } from './formModal.styles';
+import { ModalHeader } from './modalHeader/modalHeader.component';
+import { ModalBody } from './modalBody/modalBody.component';
+import { FormModalActions } from './modalButtons/modalButtons.styles';
+import { ModalCancelButton, ModalSubmitButton } from './modalButtons/modalButtons.component';
+
+export interface IFormModal extends Omit<DetailedHTMLProps<FormHTMLAttributes<HTMLFormElement>, HTMLFormElement>, 'ref'> {
+	onSubmit?: (event) => void;
+	onClickClose?: () => void;
+	onClickCancel?: () => void;
+	title?: any;
+	subtitle?: string;
+	open?: boolean;
+	confirmLabel?: string;
+	cancelLabel?: string;
+	isValid?: boolean;
+	maxWidth?: DialogProps['maxWidth'];
+	isSubmitting?: boolean;
+	disableClosing?: boolean;
+	hideHorizontalScroll?: boolean;
+}
+
+export const FormModal = ({
+	onSubmit = () => {},
+	onClickClose,
+	onClickCancel,
+	title,
+	subtitle,
+	confirmLabel,
+	cancelLabel,
+	open,
+	children,
+	className,
+	isValid = true,
+	maxWidth = false,
+	isSubmitting = false,
+	disableClosing = false,
+	hideHorizontalScroll = true,
+	...formProps
+}: IFormModal) => {
+	const handleClose = () => {
+		if (disableClosing) return;
+		(onClickCancel || onClickClose)();
```

Comment:
About the multi-step modal, is that the right behvaiour though? Say you open the `editFederationModal` and reach the second stage. When you click on `back` you sure want to go back to the first stage, but if I click on the background or even on the close button in the top bar, I think the modal should close rather than going back to the previous stage. If so, that component is what requires fixing.

About the buttons naming, the "primary" is in fact a submit button inside a form, so I would not rename that. As for the secondary one, I don't really have a strong position about that, but I would probably prefer to keep the current naming. The way I see it is if on a multi-stage form I hit "cancel", intuitively I think about cancelling what I have done on that stage in order to go back to the previous stage.
Whereas renaming it to "secondaryButton" has no actual meaning, especially if we would not rename the submitButton to "primaryButton".
Plus, in all the other modals, it is indeed a cancel button. So it feels that we are renaming a component just for an edge case. Given what the goal of this issue was, I'd prefer to create a `multiStageFormModal` if anything.

Reply:
Yes, I'm saying it's a bug how it currently works. Clicking the background or X should always close the modal. This is why the two props (onClickCancel, onClickClose) were introduced